app/geant-exporter/geant-exporter.cc

```diff
@@ -80,14 +87,176 @@ void store_particles(TFile* root_file, G4ParticleTable* particle_table)
+void loop_volumes(GeantGeometryMap& geometry, G4LogicalVolume* logical_volume)
+{
+    GeantVolume              volume;
+    GeantGeometryMap::vol_id volume_id;
+    GeantGeometryMap::mat_id material_id;
+
+    volume.name = logical_volume->GetName();
+    volume_id   = logical_volume->GetInstanceID();
+    material_id = logical_volume->GetMaterialCutsCouple()->GetIndex();
+
+    // Add volume to the global volume map
+    geometry.add_volume(volume_id, volume);
+
+    // Map volume to its material
+    geometry.link_volume_material(volume_id, material_id);
+
+    // Check for daughter volumes
+    if (logical_volume->GetNoDaughters() > 0)
+    {
+        // Repeat for every daughter volume
+        for (size_type i = 0; i < logical_volume->GetNoDaughters(); i++)
+        {
+            auto daughter_vol
+                = logical_volume->GetDaughter(i)->GetLogicalVolume();
+
+            loop_volumes(geometry, daughter_vol);
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write material table data to ROOT.
+ *
+ * The ROOT file must be open before this call.
+ */
+void store_geometry(TFile*                             root_file,
+                    G4ProductionCutsTable*             g4production_cuts,
+                    std::shared_ptr<G4VPhysicalVolume> world_volume)
+{
+    REQUIRE(root_file);
+    REQUIRE(g4production_cuts);
+    REQUIRE(world_volume);
+
+    cout << "Exporting material and volume information..." << endl;
+
+    TTree tree_materials("geometry", "geometry");
+
+    // Create geometry map and ROOT branch
+    GeantGeometryMap geometry;
+    tree_materials.Branch("GeantGeometryMap", &geometry);
+
+    // Loop over materials and elements
+    for (size_type i = 0; i < g4production_cuts->GetTableSize(); i++)
+    {
+        // Fetch material and element list
+        auto g4material_cuts = g4production_cuts->GetMaterialCutsCouple(i);
+        auto g4material      = g4material_cuts->GetMaterial();
+        auto g4elements      = g4material->GetElementVector();
+
+        // Populate material information
+        GeantMaterial material;
+        material.name             = g4material->GetName();
+        material.state            = (MaterialState)g4material->GetState();
+        material.temperature      = g4material->GetTemperature();
+        material.density          = g4material->GetDensity() / (g / cm3);
+        material.electron_density = g4material->GetTotNbOfElectPerVolume()
+                                    / (1. / cm3);
+        material.atomic_density = g4material->GetTotNbOfAtomsPerVolume()
+                                  / (1. / cm3);
+        material.radiation_length   = g4material->GetRadlen() / cm;
+        material.nuclear_int_length = g4material->GetNuclearInterLength() / cm;
+
+        // Populate element information
+        for (size_type j = 0; j < g4elements->size(); j++)
+        {
+            auto g4element = g4elements->at(j);
```

Comment:
Use `const auto&`  (or sometimes the more generic `const auto&&`); remember `auto` defaults to the value of the type, not a reference to the value.

And actually here since `g4element` is a pointer, it's making a value copy of the pointer. But most clear would be
```c++
const auto* g4element =
```
or even better
```
const auto& g4element = *g4elements->at(j);
```
since then you can access without dereferencing. If you suspect the pointer might be null, then get the pointer and put a `CHECK(g4element).`

Reply:
Done.